File: 문자열처리/[PCCP모의1차]_1)외톨이알파벳.py
# 외톨이 알파벳 = count 2회이상 && 검색그룹간 1칸 이상 떨어진 경우
# 1) Python defaultdict   2)JS정규표현식 풀이(그룹캡쳐,백레퍼런싱,) 
from collections import defaultdict
def solution(input_string):
    answer=[]
    count_table = defaultdict(list) # 딕셔너리 참조 기본값을 []로 설정
    for i,v in enumerate(input_string):
        count_table[v].append(i) # 알파벳별로 검색되는 idx를 추가;
    for key,lists in count_table.items():
        if len(lists)>=2: #2번이상 검색되는 요소만
            for i in range(len(lists)-1):
                if lists[i+1]-lists[i]>=2:
                    if key not in answer:
                        answer.append(key)
    if len(answer) == 0:
        return 'N'
    answer.sort()
    return ''.join(answer)

File: 문자열처리/[PCCP모의1차]_1)외톨이알파벳.js
//정규표현식 , string.replace(rgx,newstring)
function solution(input_string) {
  var answer = '';
  const newstr = input_string.replace(/([a-z])\1+/g,'$1')
  // regex에서 \1은 그룹을 백레퍼런스하는 기호 (replacement에서는 $ )
  console.log(newstr)
  const tempoSet = new Set()
  const ansSet = new Set()
  
  for (const a of newstr){
      if(!tempoSet.has(a)){
          tempoSet.add(a) //없으면 추가
      }
      else{//있으면 ansSet에 추가 
          ansSet.add(a)
      }
  }
  return ansSet.size ? [...ansSet].sort().join('') : 'N'
}
// set은 중복을 제거하므로 리스트로 구현시 오답이 됨.
//     const tempoSet = []
//     const ansSet = []
//     for (let a of newstr){
//         if (!tempoSet.includes(a)){
//             tempoSet.push(a) //확인배열에 없으면 정답에 추가
//         }
//         else{ //확인배열에 있으면
//             ansSet.push(a)
//         }
//     }
//     return ansSet.length ? ansSet.sort().join('') : 'N'
// }
